src/components/VolumeMeter.tsx

```diff
@@ -129,37 +132,40 @@ export const VolumeMeter = ({
     setContextState(audioContext.state);
   }, [audioContext]);
 
-  if (canvas.current) {
-    const canvasCtx = getCanvasContext(canvas.current);
-    const renderer =
-      shape === VmShape.VM_CIRCLE
-        ? new CircleRenderer(canvasCtx, {
-            width,
-            height,
-            shape,
-          })
-        : new BlockRenderer(canvasCtx, {
-            width,
-            height,
-            shape,
-            blocks,
-          });
-
-    if (anima.current) {
-      anima.current.stop();
+  useEffect(() => {
+    if (canvas.current) {
+      const canvasCtx = getCanvasContext(canvas.current);
+      rendererRef.current =
+        shape === VmShape.VM_CIRCLE
+          ? new CircleRenderer(canvasCtx, {
+              width,
+              height,
+              shape,
+            })
+          : new BlockRenderer(canvasCtx, {
+              width,
+              height,
+              shape,
+              blocks,
+            });
     }
-
-    const ani = new Animator(audioContext, renderer);
-
-    ani.changeStream(stream);
-
-    if (enabled) {
-      ani.start();
-    } else {
-      ani.stop();
+  }, []);
+  useEffect(() => {
+    if (rendererRef.current) {
+      animatorRef.current = new Animator(audioContext, rendererRef.current);
+    }
+    // intentionally not using enabled in the dependencies here - only want this updating once
+  }, [rendererRef.current]);
+  useEffect(() => {
+    if (animatorRef.current) {
+      animatorRef.current.changeStream(stream);
+      if (enabled) {
+        animatorRef.current.start();
+      } else {
+        animatorRef.current.stop();
+      }
     }
-    anima.current = ani;
-  }
+  }, [animatorRef.current, enabled, animatorRef]);
```

Comment:
```suggestion
  }, [animatorRef.current, enabled, stream]);
```

I think that's what you meant... otherwise looks great!